cmd/bancheck/bannedapi/banned_api.go

```diff
@@ -0,0 +1,136 @@
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bannedapi
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"go/token"
+	"go/types"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/go-safeweb/cmd/bancheck/config"
+	"golang.org/x/tools/go/analysis"
+)
+
+// NewAnalyzer returns an analyzer that checks for usage of banned APIs.
+func NewAnalyzer() *analysis.Analyzer {
+	fs := flag.NewFlagSet("", flag.ExitOnError)
+	fs.String("configs", "", "Config files with banned APIs separated by a comma")
+
+	a := &analysis.Analyzer{
+		Name:  "bannedAPI",
+		Doc:   "Checks for usage of banned APIs",
+		Run:   checkBannedAPIs,
+		Flags: *fs,
+	}
+
+	return a
+}
+
+func checkBannedAPIs(pass *analysis.Pass) (interface{}, error) {
+	cfgFiles := pass.Analyzer.Flags.Lookup("configs").Value.String()
+	if cfgFiles == "" {
+		return nil, errors.New("missing config files")
+	}
+
+	cfg, err := config.ReadConfigs(strings.Split(cfgFiles, ","))
+	if err != nil {
+		return nil, err
+	}
+
+	checkBannedImports(pass, bannedAPIMap(cfg.Imports))
+	checkBannedFunctions(pass, bannedAPIMap(cfg.Functions))
+
+	return nil, nil
+}
+
+func checkBannedImports(pass *analysis.Pass, bannedImports map[string][]config.BannedAPI) (interface{}, error) {
+	for _, f := range pass.Files {
+		for _, i := range f.Imports {
+			importName := strings.Trim(i.Path.Value, "\"")
+			err := reportIfBanned(importName, bannedImports, i.Pos(), pass)
+			if err != nil {
+				return false, err
+			}
+		}
+	}
+	return nil, nil
+}
+
+func checkBannedFunctions(pass *analysis.Pass, bannedFns map[string][]config.BannedAPI) (interface{}, error) {
+	for id, obj := range pass.TypesInfo.Uses {
+		fn, ok := obj.(*types.Func)
+		if !ok {
+			continue
+		}
+
+		fnName := fmt.Sprintf("%s.%s", fn.Pkg().Path(), fn.Name())
+		err := reportIfBanned(fnName, bannedFns, id.Pos(), pass)
+		if err != nil {
+			return false, err
+		}
+	}
+	return nil, nil
+}
+
+func reportIfBanned(apiName string, bannedAPIs map[string][]config.BannedAPI, position token.Pos, pass *analysis.Pass) error {
+	bannedAPICfgs, isBanned := bannedAPIs[apiName]
+	if !isBanned {
+		return nil
+	}
+	pkgAllowed, err := isPkgAllowed(pass.Pkg, bannedAPICfgs)
+	if err != nil {
+		return err
+	}
+	if pkgAllowed {
+		return nil
+	}
+	for _, bannedAPICfg := range bannedAPICfgs {
+		pass.Report(analysis.Diagnostic{
+			Pos:     position,
+			Message: fmt.Sprintf("Banned API found %q. Additional info: %s", apiName, bannedAPICfg.Msg),
+		})
+	}
+	return nil
+}
+
+// isPkgAllowed checks if the Go package should be exempted from reporting banned API usages.
+func isPkgAllowed(pkg *types.Package, bannedAPI []config.BannedAPI) (bool, error) {
```

Comment:
It might have not been from my suggestion with the two for loops nested: IMHO it's not enough to have one exemption. 

If a team adds a banned API and we end up merging configs, the API should only be allowed if it's exempted from all the checks. Note that in the future, we might end up having severities too.

All in all, I think you should just iterate over all configs that are relevant for a given API and for each do the check + report, instead of doing a check against all configs first and then do another pass and report.

Reply:
Ok, I get it now. It absolutely makes sense.

Reply:
> If a team adds a banned API and we end up merging configs, the API should only be allowed if it's exempted from all the checks.

You haven't done this, have you? https://github.com/google/go-safeweb/pull/282/files#diff-a284ab419177f0ba2e5a4a678d36212c8eddde985ef45651416a09e8964b1fffR121

This makes it so that one exemption is enough, even if there are other configs that ban it.

Reply:
Nevermind, you did. :)